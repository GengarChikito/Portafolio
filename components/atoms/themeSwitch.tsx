import { Switch, Tooltip } from "antd";
import { MoonOutlined, SunOutlined } from "@ant-design/icons";

/**
 * Defino las 'props' para mi interruptor de tema.
 * - 'dark': Un booleano que me dice si el modo oscuro está activo o no.
 * - 'onChange': La función que se va a ejecutar cuando haga clic en el switch.
 * Me devuelve el nuevo valor (true si se activa, false si se desactiva).
 */
export interface ThemeSwitchProps {
    dark: boolean;
    onChange: (value: boolean) => void;
}

/**
 * Este es mi componente 'ThemeSwitch'. Es un átomo simple
 * que uso en el Header para permitir al usuario cambiar
 * entre el modo claro y el modo oscuro.
 */
export function ThemeSwitch({ dark, onChange }: ThemeSwitchProps) {
    return (
        // Envolví el Switch en un Tooltip de Ant Design...
        // ...así, cuando el usuario pasa el mouse por encima,
        // le digo qué es lo que hará el botón (ej. "Modo claro").
        <Tooltip title={dark ? "Modo claro" : "Modo oscuro"}>

            {/* Este es el interruptor (Switch) principal de Ant Design.
              - 'checked={dark}': Le digo que esté "encendido" si la prop 'dark' es true.
              - 'onChange={onChange}': Le paso la función que recibí en las props
                                       para que la llame cuando se haga clic.
              - 'checkedChildren': Es el ícono que muestro CUANDO está encendido (Luna).
              - 'unCheckedChildren': Es el ícono CUANDO está apagado (Sol).
            */}
            <Switch
                checked={dark}
                onChange={onChange}
                checkedChildren={<MoonOutlined />}
                unCheckedChildren={<SunOutlined />}
            />
        </Tooltip>
    );
}