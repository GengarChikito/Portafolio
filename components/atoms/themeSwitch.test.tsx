import { describe, it, expect, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import React from "react";
import { ThemeSwitch } from "./themeSwitch";


// ðŸ§© Mock de Ã­conos y antd
const switchSpy = vi.fn();

vi.mock("@ant-design/icons", () => ({
    MoonOutlined: () => <span data-testid="moon-icon" />,
    SunOutlined: () => <span data-testid="sun-icon" />,
}));

vi.mock("antd", () => ({
    Tooltip: ({ children, title }: any) => (
        <div data-testid="tooltip" data-title={title}>
            {children}
        </div>
    ),
    Switch: (props: any) => {
        switchSpy(props);
        const { checked, onChange, checkedChildren, unCheckedChildren } = props;
        return (
            <button
                data-testid="switch"
                data-checked={checked}
                onClick={() => onChange(!checked)}
            >
                {checked ? checkedChildren : unCheckedChildren}
            </button>
        );
    },
}));


describe("themeSwitch", () => {
    it("muestra el tooltip con el texto correcto segÃºn el modo", () => {
        const { rerender } = render(<ThemeSwitch dark={false} onChange={() => {}} />);
        const tooltip = screen.getByTestId("tooltip");
        expect(tooltip).toHaveAttribute("data-title", "Modo oscuro");

        rerender(<ThemeSwitch dark={true} onChange={() => {}} />);
        expect(screen.getByTestId("tooltip")).toHaveAttribute("data-title", "Modo claro");
    });

    it("renderiza el Ã­cono correcto segÃºn el estado", () => {
        const { rerender } = render(<ThemeSwitch dark={false} onChange={() => {}} />);
        expect(screen.getByTestId("sun-icon")).toBeInTheDocument();
        expect(screen.queryByTestId("moon-icon")).toBeNull();

        rerender(<ThemeSwitch dark={true} onChange={() => {}} />);
        expect(screen.getByTestId("moon-icon")).toBeInTheDocument();
        expect(screen.queryByTestId("sun-icon")).toBeNull();
    });

    it("llama a onChange al hacer clic y alterna el valor", () => {
        const handleChange = vi.fn();
        render(<ThemeSwitch dark={false} onChange={handleChange} />);

        const switchBtn = screen.getByTestId("switch");
        fireEvent.click(switchBtn);

        expect(handleChange).toHaveBeenCalledTimes(1);
        expect(handleChange).toHaveBeenCalledWith(true);
    });

    it("snapshot", () => {
        const { container } = render(<ThemeSwitch dark={false} onChange={() => {}} />);
        expect(container).toMatchSnapshot();
    });
});
