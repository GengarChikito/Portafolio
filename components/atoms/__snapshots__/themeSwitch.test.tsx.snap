// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`themeSwitch > snapshot 1`] = `
<div>
  <div
    data-testid="tooltip"
    data-title="Modo oscuro"
  >
    <button
      data-checked="false"
      data-testid="switch"
    >
      <span
        data-testid="sun-icon"
      />
    </button>
  </div>
</div>
`;
